diff -r 4ae8bc883554 make/build.xml
--- a/make/build.xml	Sun Mar 27 22:15:28 2016 +0200
+++ b/make/build.xml	Sun Mar 27 22:26:53 2016 +0200
@@ -92,6 +92,7 @@
     <patternset id="jtreg.packages">
         <include name="com/sun/javatest/regtest/**"/>
         <include name="com/sun/javatest/diff/**"/>
+        <include name="java/lang/reflect/**"/>
     </patternset>
 
     <patternset id="jtreg.jdk11.files">
@@ -192,7 +193,7 @@
             debug="true" debuglevel="source,lines"
             srcdir="${src.classes.dir}"
             destdir="${build.classes.dir}"
-            classpath="${build.classes.dir}:${javatest.jar}:${junit.jar}:${testng.jar}"
+            classpath="${build.classes.dir}:${javatest.jar}:${jh.jar}:${junit.jar}:${testng.jar}"
             includeantruntime="true">
             <compilerarg line="-Xlint:all"/>
             <patternset refid="jtreg.packages"/>
diff -r 4ae8bc883554 src/share/classes/com/sun/javatest/regtest/Action.java
--- a/src/share/classes/com/sun/javatest/regtest/Action.java	Sun Mar 27 22:15:28 2016 +0200
+++ b/src/share/classes/com/sun/javatest/regtest/Action.java	Sun Mar 27 22:26:53 2016 +0200
@@ -308,6 +308,12 @@
      * </pre>
      */
     protected void startAction() {
+        ActionCallBack callBack = script.getCallBack();
+
+        if (callBack != null) {
+            callBack.actionStarted(this);
+        }
+
         String name = getName();
         section = script.getTestResult().createSection(name);
 
@@ -332,6 +338,12 @@
         pw.println(LOG_ELAPSED_TIME + ((double) elapsedTime/1000.0));
         recorder.close();
         section.setStatus(status);
+
+        ActionCallBack callBack = script.getCallBack();
+
+        if (callBack != null) {
+            callBack.actionEnded(this);
+        }
     } // endAction()
 
     //----------workarounds-------------------------------------------------------
diff -r 4ae8bc883554 src/share/classes/com/sun/javatest/regtest/ActionCallBack.java
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/src/share/classes/com/sun/javatest/regtest/ActionCallBack.java	Sun Mar 27 22:26:53 2016 +0200
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2014, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.javatest.regtest;
+
+import java.util.Collection;
+
+public interface ActionCallBack {
+
+    public void actionStarted(Action action);
+
+    public Collection<? extends String> getAdditionalVMJavaOptions(Action action);
+
+    public void actionEnded(Action action);
+
+}
diff -r 4ae8bc883554 src/share/classes/com/sun/javatest/regtest/CompileAction.java
--- a/src/share/classes/com/sun/javatest/regtest/CompileAction.java	Sun Mar 27 22:15:28 2016 +0200
+++ b/src/share/classes/com/sun/javatest/regtest/CompileAction.java	Sun Mar 27 22:26:53 2016 +0200
@@ -501,6 +501,12 @@
         String javacCmd = script.getJavacProg();
 
         List<String> javacVMOpts = new ArrayList<String>();
+        ActionCallBack callBack = script.getCallBack();
+
+        if (callBack != null) {
+            javacVMOpts.addAll(callBack.getAdditionalVMJavaOptions(this));
+        }
+
         javacVMOpts.addAll(script.getTestVMOptions());
         if (addDebugOpts && script.getCompileJDK().equals(script.getTestJDK()))
             javacVMOpts.addAll(script.getTestDebugOptions());
diff -r 4ae8bc883554 src/share/classes/com/sun/javatest/regtest/Main.java
--- a/src/share/classes/com/sun/javatest/regtest/Main.java	Sun Mar 27 22:15:28 2016 +0200
+++ b/src/share/classes/com/sun/javatest/regtest/Main.java	Sun Mar 27 22:26:53 2016 +0200
@@ -1050,6 +1050,8 @@
         return run();
     }
 
+    public static ActionCallBack callBack;
+
     private int run() throws BadArgs, Fault, Harness.Fault, InterruptedException {
         findSystemJarFiles();
 
@@ -1102,10 +1104,10 @@
             Version requiredVersion = ts.getRequiredVersion();
             Version currentVersion = Version.getCurrent();
             if (requiredVersion.compareTo(currentVersion) > 0) {
-                throw new Fault(i18n, "main.requiredVersion",
-                        ts.getPath(),
-                        requiredVersion.version, requiredVersion.build,
-                        currentVersion.version, currentVersion.build);
+                err.println(i18n.getString("main.requiredVersion", new Object[] {
+                                           ts.getPath(),
+                                           requiredVersion.version, requiredVersion.build,
+                                           currentVersion.version, currentVersion.build}));
             }
         }
 
@@ -1781,6 +1783,8 @@
 
             rp.initExprContext(); // will invoke/init jdk.getProperties(params)
 
+            rp.setCallBack(callBack);
+
             return rp;
         } catch (TestSuite.Fault f) {
             // TODO: fix bad string -- need more helpful resource here
diff -r 4ae8bc883554 src/share/classes/com/sun/javatest/regtest/MainAction.java
--- a/src/share/classes/com/sun/javatest/regtest/MainAction.java	Sun Mar 27 22:15:28 2016 +0200
+++ b/src/share/classes/com/sun/javatest/regtest/MainAction.java	Sun Mar 27 22:26:53 2016 +0200
@@ -413,6 +413,12 @@
         javaOpts.addAll(script.getTestVMJavaOptions());
         javaOpts.addAll(script.getTestDebugOptions());
 
+        ActionCallBack callBack = script.getCallBack();
+
+        if (callBack != null) {
+            javaOpts.addAll(new ArrayList<String>(callBack.getAdditionalVMJavaOptions(this)));
+        }
+
         Map<String, String> javaProps = new LinkedHashMap<String, String>();
         javaProps.putAll(script.getTestProperties());
 
diff -r 4ae8bc883554 src/share/classes/com/sun/javatest/regtest/RegressionParameters.java
--- a/src/share/classes/com/sun/javatest/regtest/RegressionParameters.java	Sun Mar 27 22:15:28 2016 +0200
+++ b/src/share/classes/com/sun/javatest/regtest/RegressionParameters.java	Sun Mar 27 22:26:53 2016 +0200
@@ -1110,4 +1110,14 @@
     private final Set<Integer> retainStatusSet = new HashSet<Integer>(4);
     private Pattern retainFilesPattern;
 
+    private ActionCallBack callBack;
+
+    public ActionCallBack getCallBack() {
+        return callBack;
+    }
+
+    public void setCallBack(ActionCallBack callBack) {
+        this.callBack = callBack;
+    }
+
 }
diff -r 4ae8bc883554 src/share/classes/com/sun/javatest/regtest/RegressionScript.java
--- a/src/share/classes/com/sun/javatest/regtest/RegressionScript.java	Sun Mar 27 22:15:28 2016 +0200
+++ b/src/share/classes/com/sun/javatest/regtest/RegressionScript.java	Sun Mar 27 22:26:53 2016 +0200
@@ -993,6 +993,10 @@
         return params.getCompileJDK().getJavacProg().getPath();
     }
 
+    ActionCallBack getCallBack() {
+        return params.getCallBack();
+    }
+
     //--------------------------------------------------------------------------
 
     // Get the standard properties to be set for tests
diff -r 58c0d0e99a17 src/share/classes/com/sun/javatest/regtest/JDK.java
--- a/src/share/classes/com/sun/javatest/regtest/JDK.java	Wed Jul 06 13:35:18 2016 +0200
+++ b/src/share/classes/com/sun/javatest/regtest/JDK.java	Wed Jul 06 13:38:50 2016 +0200
@@ -80,6 +80,10 @@
 
     private static final Map<File, JDK> cache = new HashMap<File, JDK>();
 
+    public static synchronized void clearCache() {
+        cache.clear();
+    }
+
     private JDK(File jdk) {
         this.jdk = jdk;
         absJDK = jdk.getAbsoluteFile();
